### PR TITLE
ATO-554 add BlockingIOError handler, tests and portalocker

### PR DESCRIPTION
**Pull request by Urkem**
**Proposed changes**:
- Fix [ATO-554](https://rasahq.atlassian.net/browse/ATO-554) by obtaining the STDOUT lock with portalocker, as it blocks only half way on utterances with more than ~5KB of text.

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

[ATO-554]: https://rasahq.atlassian.net/browse/ATO-554?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: changelog/12271.bugfix.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix `BlockingIOError` when running `rasa cli` on utterances with more than 5KB of text.
```
Comment (ancalita) [inline changelog/12271.bugfix.md:1]: We should be more specific about which CLI command causes this issue (I think it's `rasa shell`, could also be `rasa interactive`), please double-check with @trammell.

File: pyproject.toml [hunk 2]
```diff
@@ -148,6 +148,7 @@ randomname = "^0.1.5"
 pluggy = "^1.0.0"
 slack-sdk = "^3.19.2"
 confluent-kafka = "^1.9.2"
+portalocker = "^2.7.0"
```
Comment (ancalita) [inline pyproject.toml:151]: Please double-check that the license of this new dependency is compatible with our use-case(s) in rasa.

File: rasa/shared/utils/cli.py [hunk 3]
```diff
@@ -6,15 +6,17 @@
...
-    try:
+    stream = sys.stdout
+    if sys.platform == "win32":
         # colorama is used to fix a regression where colors can not be printed on
         # windows. https://github.com/RasaHQ/rasa/issues/7053
         from colorama import AnsiToWin32
 
         stream = AnsiToWin32(sys.stdout).stream
+    try:
         print(output, file=stream)
```
Comment (ancalita) [inline rasa/shared/utils/cli.py:17]: Do we actually need the `stream` variable here for all platforms? In the original code it seems to be valid only for Windows 🤔

File: tests/shared/utils/test_io.py [hunk 4]
```diff
@@ -620,3 +623,22 @@ def test_deep_container_fingerprint_can_use_instance_fingerprint():
...
+
+# Can't manipulate the STDOUT file descriptor
+# in pytest (it's a io.StringIO instance not a file)
+def test_handle_print_blocking(monkeypatch: MonkeyPatch):
+    def mock_print(*args, **kwargs):
+        raise BlockingIOError()
+
+    mock = Mock()
+    monkeypatch.setattr(builtins, "print", mock_print)
+    monkeypatch.setattr(rasa.shared.utils.io, "handle_print_blocking", mock)
```
Comment (ancalita) [inline tests/shared/utils/test_io.py:636]: If we're mocking the function we want to test, we're not actually testing it, but I'd suggest to at least assert that mock was called with the desired output (in addition to the `call_count` assertion).
Also pytest has a good fixture for capturing stdout: `capsys` (https://docs.pytest.org/en/7.1.x/how-to/capture-stdout-stderr.html), have you tried with that too?

**Review by ancalita — COMMENTED** [review]
Looks good 👍🏻 
Left some suggestions in the comments.

File: tests/shared/utils/test_io.py (hunk 4 above)
Comment (Urkem) [reply tests/shared/utils/test_io.py:636]: Tried with `capsys` but it captures only the text output to stdout/stderr and not the system calls to lock/unlock stdout. After trying with `capsys` and `capfd` I decided to mock it but there was no good way (that I know of) to assert that lock/unlock happened so I was left with this...

File: rasa/shared/utils/cli.py (hunk 3 above)
Comment (Urkem) [reply rasa/shared/utils/cli.py:17]: By default `print` gets called with  `end='\n', file=sys.stdout, flush=False` I just made it explicit to not have the two `print` statements as before

File: rasa/shared/utils/io.py [hunk 5]
```diff
@@ -626,3 +625,21 @@ def is_subdirectory(path: Text, potential_parent_directory: Text) -> bool:
 def random_string(length: int) -> Text:
     """Returns a random string of given length."""
     return "".join(random.choices(string.ascii_uppercase + string.digits, k=length))
+
+
+def handle_print_blocking(output: Text) -> None:
```
Comment (radovanZRasa) [inline rasa/shared/utils/io.py:630]: We must add unit test for this function to make sure that:
* portalocker.Lock is called properly - mock `Lock` function
* AnsiToWin32 is called properly if `sys.platform == "win32"`
* print is called with proper `output`, `file` and `flush` parameters

**Review by radovanZRasa — CHANGES_REQUESTED** [review]

**Comment by radovanZRasa** [comment]
One Note:
If we were able to reproduce the issue, why couldn't we write the test which reproduces it?

**Comment by Urkem** [comment]
@radovanZRasa I was not able to reproduce it 100% of the time It happens mostly on MacOS `python 3.8` 80% of the time and on `3.7 3.9 3.10` I was able to reproduce it 5-10% of the time. On Windows I could not reproduce it (I raised the exception explicitly)... @trammell stated in the jira ticket he is not able to reproduce it at will

File: pyproject.toml (hunk 2 above)
Comment (Urkem) [reply pyproject.toml:151]: Its the same license as [numpy](https://github.com/numpy/numpy/blob/main/LICENSE.txt) (BSD-3) so I think its fine 😃

Comment (ancalita) [reply pyproject.toml:151]: Perfect!

**Review by radovanZRasa — APPROVED** [review]

**Pull request merged**